File: tasks/translation/output/Qwen/Qwen2.5-32B-Instruct-AWQ/codenet/direct/Java/Python/s514557145_comments_removed.py
MOD2 = 998244353

class S:
    E = None

    def __init__(self, sum, l, r):
        self.sum = sum
        self.l = l
        self.r = r

    @staticmethod
    def op(s1, s2):
        sum = (s1.sum + s2.sum) % MOD2
        return S(sum, min(s1.l, s2.l), max(s1.r, s2.r))

    @staticmethod
    def map(f, s):
        c = (tenmod[f.N - s.l] - tenmod[f.N - s.r - 1]) % MOD2
        if c < 0:
            c += MOD2
        return S(((f.D * c) % MOD2 * modInv(9)) % MOD2, s.l, s.r)

class F:
    I = None

    def __init__(self, D, N):
        self.D = D
        self.N = N

    @staticmethod
    def composite(f, g):
        return F(f.D, f.N)

def modInv(x):
    return modPow(x, MOD2 - 2)

def modPow(x, y):
    z = 1
    while y > 0:
        if y % 2 == 0:
            x = (x * x) % MOD2
            y //= 2
        else:
            z = (z * x) % MOD2
            y -= 1
    return z

class LazySegTree:
    def __init__(self, dat, op, E, mapping, composition, Id):
        self.MAX = len(dat)
        self.N = 1
        while self.N < self.MAX:
            self.N <<= 1
        self.Log = self.N.bit_length() - 1
        self.Op = op
        self.E = E
        self.Mapping = mapping
        self.Composition = composition
        self.Id = Id
        self.Dat = [E] * (self.N << 1)
        self.Laz = [Id] * self.N
        self.build(dat)

    def build(self, dat):
        l = len(dat)
        self.Dat[self.N:self.N + l] = dat
        for i in range(self.N - 1, 0, -1):
            self.Dat[i] = self.Op(self.Dat[i << 1 | 0], self.Dat[i << 1 | 1])

    def push(self, k):
        if self.Laz[k] == self.Id:
            return
        lk, rk = k << 1 | 0, k << 1 | 1
        self.Dat[lk] = self.Mapping(self.Laz[k], self.Dat[lk])
        self.Dat[rk] = self.Mapping(self.Laz[k], self.Dat[rk])
        if lk < self.N:
            self.Laz[lk] = self.Composition(self.Laz[k], self.Laz[lk])
        if rk < self.N:
            self.Laz[rk] = self.Composition(self.Laz[k], self.Laz[rk])
        self.Laz[k] = self.Id

    def pushTo(self, k):
        for i in range(self.Log, 0, -1):
            self.push(k >> i)

    def pushToRange(self, lk, rk):
        for i in range(self.Log, 0, -1):
            if ((lk >> i) << i) != lk:
                self.push(lk >> i)
            if ((rk >> i) << i) != rk:
                self.push(rk >> i)

    def updateFrom(self, k):
        k >>= 1
        while k > 0:
            self.Dat[k] = self.Op(self.Dat[k << 1 | 0], self.Dat[k << 1 | 1])
            k >>= 1

    def updateFromRange(self, lk, rk):
        for i in range(1, self.Log + 1):
            if ((lk >> i) << i) != lk:
                lki = lk >> i
                self.Dat[lki] = self.Op(self.Dat[lki << 1 | 0], self.Dat[lki << 1 | 1])
            if ((rk >> i) << i) != rk:
                rki = (rk - 1) >> i
                self.Dat[rki] = self.Op(self.Dat[rki << 1 | 0], self.Dat[rki << 1 | 1])

    def set(self, p, x):
        self.exclusiveRangeCheck(p)
        p += self.N
        self.pushTo(p)
        self.Dat[p] = x
        self.updateFrom(p)

    def get(self, p):
        self.exclusiveRangeCheck(p)
        p += self.N
        self.pushTo(p)
        return self.Dat[p]

    def allProd(self):
        return self.Dat[1]

    def apply(self, p, f):
        self.exclusiveRangeCheck(p)
        p += self.N
        self.pushTo(p)
        self.Dat[p] = self.Mapping(f, self.Dat[p])
        self.updateFrom(p)

    def applyRange(self, l, r, f):
        if l > r:
            raise ValueError(f"Invalid range: [{l}, {r})")
        self.inclusiveRangeCheck(l)
        self.inclusiveRangeCheck(r)
        if l == r:
            return
        l += self.N
        r += self.N
        self.pushToRange(l, r)
        while l < r:
            if l & 1:
                self.Dat[l] = self.Mapping(f, self.Dat[l])
                if l < self.N:
                    self.Laz[l] = self.Composition(f, self.Laz[l])
                l += 1
            if r & 1:
                r -= 1
                self.Dat[r] = self.Mapping(f, self.Dat[r])
                if r < self.N:
                    self.Laz[r] = self.Composition(f, self.Laz[r])
            l >>= 1
            r >>= 1
        self.updateFromRange(l, r)

    def maxRight(self, l, g):
        self.inclusiveRangeCheck(l)
        if not g(self.E):
            raise ValueError("Identity element must satisfy the condition.")
        if l == self.MAX:
            return self.MAX
        l += self.N
        self.pushTo(l)
        sum = self.E
        while True:
            l >>= l.bit_length() - 1
            if not g(self.Op(sum, self.Dat[l])):
                while l < self.N:
                    self.push(l)
                    l <<= 1
                    if g(self.Op(sum, self.Dat[l])):
                        sum = self.Op(sum, self.Dat[l])
                        l += 1
                return l - self.N
            sum = self.Op(sum, self.Dat[l])
            l += 1
            if (l & -l) == l:
                break
        return self.MAX

    def minLeft(self, r, g):
        self.inclusiveRangeCheck(r)
        if not g(self.E):
            raise ValueError("Identity element must satisfy the condition.")
        if r == 0:
            return 0
        r += self.N
        self.pushTo(r - 1)
        sum = self.E
        while True:
            r -= 1
            while r > 1 and (r & 1) == 1:
                r >>= 1
            if not g(self.Op(self.Dat[r], sum)):
                while r < self.N:
                    self.push(r)
                    r = r << 1 | 1
                    if g(self.Op(self.Dat[r], sum)):
                        sum = self.Op(self.Dat[r], sum)
                        r -= 1
                return r + 1 - self.N
            sum = self.Op(self.Dat[r], sum)
            if (r & -r) == r:
                break
        return 0

    def exclusiveRangeCheck(self, p):
        if p < 0 or p >= self.MAX:
            raise IndexError(f"Index {p} is not in [{0}, {self.MAX}).")

    def inclusiveRangeCheck(self, p):
        if p < 0 or p > self.MAX:
            raise IndexError(f"Index {p} is not in [{0}, {self.MAX}].")

    def setIndent(self, newIndent):
        self.indent = newIndent

    def __str__(self):
        return self.toString(1, 0)

    def toString(self, k, sp):
        if k >= self.N:
            return self.indent(sp) + str(self.Dat[k])
        s = self.toString(k << 1 | 1, sp + self.indent)
        s += "\n"
        s += self.indent(sp) + str(self.Dat[k]) + "/" + str(self.Laz[k])
        s += "\n"
        s += self.toString(k << 1 | 0, sp + self.indent)
        return s

    def indent(self, n):
        return ' ' * n

class InputReader:
    def __init__(self, in_stream):
        self.in_stream = in_stream
        self.buffer = bytearray(1024)
        self.curbuf = self.lenbuf = 0

    def hasNextByte(self):
        if self.curbuf >= self.lenbuf:
            self.curbuf = 0
            try:
                self.lenbuf = self.in_stream.readinto(self.buffer)
            except Exception:
                raise EOFError("Input stream has ended unexpectedly.")
            if self.lenbuf <= 0:
                return False
        return True

    def readByte(self):
        if self.hasNextByte():
            b = self.buffer[self.curbuf]
            self.curbuf += 1
            return b
        else:
            return -1

    def isSpaceChar(self, c):
        return not (33 <= c <= 126)

    def skip(self):
        while self.hasNextByte() and self.isSpaceChar(self.buffer[self.curbuf]):
            self.curbuf += 1

    def hasNext(self):
        self.skip()
        return self.hasNextByte()

    def next(self):
        if not self.hasNext():
            raise EOFError("No more input available.")
        sb = bytearray()
        b = self.readByte()
        while not self.isSpaceChar(b):
            sb.append(b)
            b = self.readByte()
        return sb.decode()

    def nextInt(self):
        if not self.hasNext():
            raise EOFError("No more input available.")
        c = self.readByte()
        while self.isSpaceChar(c):
            c = self.readByte()
        minus = False
        if c == ord('-'):
            minus = True
            c = self.readByte()
        res = 0
        while True:
            if not ('0' <= chr(c) <= '9'):
                raise ValueError("Invalid input for integer.")
            res = res * 10 + c - ord('0')
            c = self.readByte()
            if self.isSpaceChar(c):
                break
        return -res if minus else res

    def nextLong(self):
        if not self.hasNext():
            raise EOFError("No more input available.")
        c = self.readByte()
        while self.isSpaceChar(c):
            c = self.readByte()
        minus = False
        if c == ord('-'):
            minus = True
            c = self.readByte()
        res = 0
        while True:
            if not ('0' <= chr(c) <= '9'):
                raise ValueError("Invalid input for long.")
            res = res * 10 + c - ord('0')
            c = self.readByte()
            if self.isSpaceChar(c):
                break
        return -res if minus else res

    def nextDouble(self):
        return float(self.next())

    def nextIntArray(self, n):
        return [self.nextInt() for _ in range(n)]

    def nextLongArray(self, n):
        return [self.nextLong() for _ in range(n)]

    def nextCharMap(self, n, m):
        return [self.next().encode() for _ in range(n)]

def main():
    out = sys.stdout
    sc = InputReader(sys.stdin.buffer)
    N = sc.nextInt()
    Q = sc.nextInt()
    global tenmod
    tenmod = [1] * (N + 1)
    for j in range(1, N + 1):
        tenmod[j] = (tenmod[j - 1] * 10) % MOD2
    S.E = S(0, float('inf') // 2, -float('inf') // 2)
    F.I = F(0, 0)
    dat = [S(tenmod[N - i - 1], i, i) for i in range(N)]
    lazySegTree = LazySegTree(dat, S.op, S.E, S.map, F.composite, F.I)
    for _ in range(Q):
        l = sc.nextInt() - 1
        r = sc.nextInt() - 1
        D = sc.nextLong()
        lazySegTree.applyRange(l, r + 1, F(D, N))
        out.write(str(lazySegTree.allProd().sum) + '\n')
    out.flush()

if __name__ == "__main__":
    import sys
    main()
# End of Code

File: tasks/translation/output/Qwen/Qwen2.5-32B-Instruct-AWQ/codenet/direct/Java/C/s514557145_comments_removed.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MOD2 998244353

typedef struct {
    long long sum;
    int l;
    int r;
} S;

typedef struct {
    long long D;
    int N;
} F;

S S_E = {0, INT_MAX / 2, INT_MIN / 2};
F F_I = {0, 0};

long long modInv(long long x) {
    return modPow(x, MOD2 - 2);
}

long long modPow(long long x, long long y) {
    long long z = 1;
    while (y > 0) {
        if (y % 2 == 0) {
            x = (x * x) % MOD2;
            y /= 2;
        } else {
            z = (z * x) % MOD2;
            y--;
        }
    }
    return z;
}

S S_op(S s1, S s2) {
    long long sum = s1.sum + s2.sum;
    if (sum >= MOD2) {
        sum -= MOD2;
    }
    return (S){sum, (s1.l < s2.l) ? s1.l : s2.l, (s1.r > s2.r) ? s1.r : s2.r};
}

S S_map(F f, S s) {
    long long c = (tenmod[f.N - s.l] - tenmod[f.N - s.r - 1]);
    if (c < 0) {
        c += MOD2;
    }
    return (S){((f.D * c) % MOD2 * modInv(9)) % MOD2, s.l, s.r};
}

F F_composite(F f, F g) {
    return (F){f.D, f.N};
}

void exclusiveRangeCheck(int p, int MAX) {
    if (p < 0 || p >= MAX) {
        fprintf(stderr, "Index %d is not in [%d, %d).\n", p, 0, MAX);
        exit(1);
    }
}

void inclusiveRangeCheck(int p, int MAX) {
    if (p < 0 || p > MAX) {
        fprintf(stderr, "Index %d is not in [%d, %d].\n", p, 0, MAX);
        exit(1);
    }
}

void push(int k, S *Dat, F *Laz, int N, int Log) {
    if (Laz[k] == F_I) return;
    int lk = k << 1 | 0, rk = k << 1 | 1;
    Dat[lk] = S_map(Laz[k], Dat[lk]);
    Dat[rk] = S_map(Laz[k], Dat[rk]);
    if (lk < N) Laz[lk] = F_composite(Laz[k], Laz[lk]);
    if (rk < N) Laz[rk] = F_composite(Laz[k], Laz[rk]);
    Laz[k] = F_I;
}

void pushTo(int k, int N, int Log, S *Dat, F *Laz) {
    for (int i = Log; i > 0; i--) push(k >> i, Dat, Laz, N, Log);
}

void pushTo(int lk, int rk, int N, int Log, S *Dat, F *Laz) {
    for (int i = Log; i > 0; i--) {
        if (((lk >> i) << i) != lk) push(lk >> i, Dat, Laz, N, Log);
        if (((rk >> i) << i) != rk) push(rk >> i, Dat, Laz, N, Log);
    }
}

void updateFrom(int k, S *Dat, int N) {
    k >>= 1;
    while (k > 0) {
        Dat[k] = S_op(Dat[k << 1 | 0], Dat[k << 1 | 1]);
        k >>= 1;
    }
}

void updateFrom(int lk, int rk, S *Dat, int N) {
    for (int i = 1; i <= Log; i++) {
        if (((lk >> i) << i) != lk) {
            int lki = lk >> i;
            Dat[lki] = S_op(Dat[lki << 1 | 0], Dat[lki << 1 | 1]);
        }
        if (((rk >> i) << i) != rk) {
            int rki = (rk - 1) >> i;
            Dat[rki] = S_op(Dat[rki << 1 | 0], Dat[rki << 1 | 1]);
        }
    }
}

void set(int p, S x, S *Dat, int N) {
    exclusiveRangeCheck(p, N);
    p += N;
    pushTo(p, N, Log, Dat, Laz);
    Dat[p] = x;
    updateFrom(p, Dat, N);
}

S get(int p, S *Dat, int N) {
    exclusiveRangeCheck(p, N);
    p += N;
    pushTo(p, N, Log, Dat, Laz);
    return Dat[p];
}

S allProd(S *Dat) {
    return Dat[1];
}

void apply(int p, F f, S *Dat, F *Laz, int N) {
    exclusiveRangeCheck(p, N);
    p += N;
    pushTo(p, N, Log, Dat, Laz);
    Dat[p] = S_map(f, Dat[p]);
    updateFrom(p, Dat, N);
}

void apply(int l, int r, F f, S *Dat, F *Laz, int N, int Log) {
    if (l > r) {
        fprintf(stderr, "Invalid range: [%d, %d)\n", l, r);
        exit(1);
    }
    inclusiveRangeCheck(l, N);
    inclusiveRangeCheck(r, N);
    if (l == r) return;
    l += N; r += N;
    pushTo(l, r, N, Log, Dat, Laz);
    for (int l2 = l, r2 = r; l2 < r2;) {
        if ((l2 & 1) == 1) {
            Dat[l2] = S_map(f, Dat[l2]);
            if (l2 < N) Laz[l2] = F_composite(f, Laz[l2]);
            l2++;
        }
        if ((r2 & 1) == 1) {
            r2--;
            Dat[r2] = S_map(f, Dat[r2]);
            if (r2 < N) Laz[r2] = F_composite(f, Laz[r2]);
        }
        l2 >>= 1; r2 >>= 1;
    }
    updateFrom(l, r, Dat, N);
}

int maxRight(int l, S *Dat, int N, int Log, F *Laz, int (*g)(S)) {
    inclusiveRangeCheck(l, N);
    if (!g(S_E)) {
        fprintf(stderr, "Identity element must satisfy the condition.\n");
        exit(1);
    }
    if (l == N) return N;
    l += N;
    pushTo(l, N, Log, Dat, Laz);
    S sum = S_E;
    do {
        l >>= __builtin_ctz(l);
        if (!g(S_op(sum, Dat[l]))) {
            while (l < N) {
                push(l, Dat, Laz, N, Log);
                l = l << 1;
                if (g(S_op(sum, Dat[l]))) {
                    sum = S_op(sum, Dat[l]);
                    l++;
                }
            }
            return l - N;
        }
        sum = S_op(sum, Dat[l]);
        l++;
    } while ((l & -l) != l);
    return N;
}

int minLeft(int r, S *Dat, int N, int Log, F *Laz, int (*g)(S)) {
    inclusiveRangeCheck(r, N);
    if (!g(S_E)) {
        fprintf(stderr, "Identity element must satisfy the condition.\n");
        exit(1);
    }
    if (r == 0) return 0;
    r += N;
    pushTo(r - 1, N, Log, Dat, Laz);
    S sum = S_E;
    do {
        r--;
        while (r > 1 && (r & 1) == 1) r >>= 1;
        if (!g(S_op(Dat[r], sum))) {
            while (r < N) {
                push(r, Dat, Laz, N, Log);
                r = r << 1 | 1;
                if (g(S_op(Dat[r], sum))) {
                    sum = S_op(Dat[r], sum);
                    r--;
                }
            }
            return r + 1 - N;
        }
        sum = S_op(Dat[r], sum);
    } while ((r & -r) != r);
    return 0;
}

int main() {
    int N, Q;
    scanf("%d %d", &N, &Q);
    long long *tenmod = (long long *)malloc((N + 1) * sizeof(long long));
    tenmod[0] = 1;
    for (int j = 1; j <= N; j++) {
        tenmod[j] = (tenmod[j - 1] * 10LL) % MOD2;
    }
    S *dat = (S *)malloc(N * sizeof(S));
    for (int j = 0; j < N; j++) {
        dat[j] = (S){tenmod[N - j - 1], j, j};
    }
    S *Dat = (S *)malloc((N << 1) * sizeof(S));
    F *Laz = (F *)malloc(N * sizeof(F));
    for (int i = 0; i < N << 1; i++) {
        Dat[i] = S_E;
    }
    for (int i = 0; i < N; i++) {
        Laz[i] = F_I;
    }
    for (int i = N - 1; i > 0; i--) {
        Dat[i] = S_op(Dat[i << 1 | 0], Dat[i << 1 | 1]);
    }
    int Log = __builtin_ctz(N);
    for (int j = 0; j < Q; j++) {
        int l, r;
        long long D;
        scanf("%d %d %lld", &l, &r, &D);
        l--; r--;
        F f = {D, N};
        apply(l, r + 1, f, Dat, Laz, N, Log);
        printf("%lld\n", allProd(Dat).sum);
    }
    free(tenmod);
    free(dat);
    free(Dat);
    free(Laz);
    return 0;
}
// End of Code
